man8/era_invalidate.8
.TH ERA_INVALIDATE 8 "Thin Provisioning Tools" "Red Hat, Inc." \" -*- nroff -*-
.SH NAME
era_invalidate \- Provide a list of blocks that have changed since a particular era.

.SH SYNOPSIS
.B era_invalidate
.RB [ options ]
.I {device|file}

.SH DESCRIPTION
.B era_invalidate
Examines era metadata and lists blocks that may have changed since a given era.

This tool cannot be run on live metadata unless the \fB\-\-metadata\-snap\fP option is used.

.SH OPTIONS
.IP "\fB\-h, \-\-help\fP"
Print help and exit.

.IP "\fB\-V, \-\-version\fP"
Output version information and exit.

.IP "\fB\-o <output file>\fP"
Write output to a file rather than
.B stdout
.

.IP "\fB\-\-metadata\-snapshot\fP"
Use the metadata snapshot rather than the current superblock.

.IP "\fB\-\-written-since {era}\fP"
Blocks written since the given era will be listed.

.SH EXAMPLE
List the blocks that may have been written since the beginning of era
13 on the metadata device /dev/vg/metadata.
.sp
.B era_invalidate --written-since 13 /dev/vg/metadata

The device may not be actively used by the target
when running.

.SH DIAGNOSTICS
.B era_invalidate
returns an exit code of 0 for success or 1 for error (eg, metadata corruption).

.SH SEE ALSO
.B era_check(8),
.B era_dump(8),
.B era_repair(8),
.B era_restore(8)

.SH AUTHOR
Joe Thornber <ejt@redhat.com>


CHANGES
v0.6.2
======

- Fix bug in thin_delta
- Fix recent regression in thin_repair.
- Force g++-98 dialect
- Fix bug in thin_trim

v0.6.1
======

Fix bug in thin_dump when using metadata snaps.

v0.6.0
======

thin_ls

v0.5.6
======

era_invalidate may be run on live metadata if the --metadata-snap option is given.

v0.5.5
======

You may now give the --metadata_snap option to thin_delta without specifying where the snap is.

Update man pages to make it clearer that most tools shouldn't be run on live metadata.

Fix some bugs in the metadata reference counting for thin_check.

v0.5.1
======

Fix bug where the tools would crash if given a very large metadata
device to restore to.

v0.5
====

- thin_delta, thin_trim
- --clear-needs-check flag for cache_check
- space map checking for thin check

v0.4
====

- All tools switch to using libaio.  This gives a large performance
  boost, especially to the write focused tools like thin_restore.

- Added a progress monitor to thin_restore, cache_restore and era_restore

- Added a --quiet/-q option to *_restore to turn off the progress bar

- Removed variable hint size support from cache tools.  The kernel
  still only supports a fixed 32bit width.  This will have a side
  effect of reducing the executable sizes due to less template
  instatiation.

- Tools rolled into a single executable to save space.

- Fixed some bugs when walking bitsets (possibly effecting cache_dump
  and cache_check).

era/era_invalidate.cc
#include <fstream>
#include <getopt.h>
#include <libgen.h>
#include <iostream>

#include "version.h"
#include "base/indented_stream.h"
#include "era/commands.h"
#include "era/era_array.h"
#include "era/writeset_tree.h"
#include "era/metadata.h"
#include "era/xml_format.h"
#include "persistent-data/file_utils.h"

#include <boost/lexical_cast.hpp>

using namespace boost;
using namespace era;
using namespace std;

//----------------------------------------------------------------

namespace {
	struct flags {
		flags()
			: metadata_snapshot_(false) {
		}

		bool metadata_snapshot_;
		optional<uint32_t> era_threshold_;
	};

	//--------------------------------

	void walk_array(era_array const &array, uint32_t nr_blocks,
			uint32_t threshold, set<uint32_t> &blocks) {
		for (uint32_t b = 0; b < nr_blocks; b++) {
			uint32_t era = array.get(b);
			if (era >= threshold)
				blocks.insert(b);
		}
	}

	class writesets_marked_since : public writeset_tree_detail::writeset_visitor {
	public:
		writesets_marked_since(uint32_t threshold, set<uint32_t> &blocks)
			: current_era_(0),
			  threshold_(threshold),
			  blocks_(blocks) {
		}

		void writeset_begin(uint32_t era, uint32_t nr_bits) {
			current_era_ = era;
		}

		void bit(uint32_t index, bool value) {
			if (value && current_era_ >= threshold_)
				blocks_.insert(index);
		}

		void writeset_end() {
		}

	private:
		uint32_t current_era_;
		uint32_t threshold_;
		set<uint32_t> &blocks_;
	};

	void raise_metadata_damage() {
		throw std::runtime_error("metadata contains errors (run era_check for details).");
	}

	struct fatal_writeset_tree_damage : public writeset_tree_detail::damage_visitor {
		void visit(writeset_tree_detail::missing_eras const &d) {
			raise_metadata_damage();
		}

		void visit(writeset_tree_detail::damaged_writeset const &d) {
			raise_metadata_damage();
		}
	};

	void walk_writesets(metadata const &md, uint32_t threshold, set<uint32_t> &result) {
		writesets_marked_since v(threshold, result);
		fatal_writeset_tree_damage dv;

		walk_writeset_tree(md.tm_, *md.writeset_tree_, v, dv);
	}

	void mark_blocks_since(metadata const &md, optional<uint32_t> const &threshold, set<uint32_t> &result) {
		if (!threshold)
			// Can't get here, just putting in to pacify the compiler
			throw std::runtime_error("threshold not set");
		else {
			walk_array(*md.era_array_, md.sb_.nr_blocks, *threshold, result);
			walk_writesets(md, *threshold, result);
		}
	}

	//--------------------------------

	template <typename Iterator>
	pair<uint32_t, uint32_t> next_run(Iterator &it, Iterator end) {
		uint32_t b, e;

		b = *it++;
		e = b + 1;
		while (it != end && *it == e) {
			e++;
			it++;
		}

		return make_pair(b, e);
	}

	void emit_blocks(ostream &out, set<uint32_t> const &blocks) {
		indented_stream o(out);

		o.indent();
		o << "<blocks>" << endl;

		o.inc();
		{
			set<uint32_t>::const_iterator it = blocks.begin();
			while (it != blocks.end()) {
				o.indent();

				pair<uint32_t, uint32_t> range = next_run(it, blocks.end());
				if (range.second - range.first == 1)
					o << "<block block=\"" << range.first << "\"/>" << endl;

				else
					o << "<range begin=\"" << range.first
					  << "\" end = \"" << range.second << "\"/>" << endl;
			}
		}
		o.dec();

		o.indent();
		o << "</blocks>" << endl;
	}

	//--------------------------------

	string const STDOUT_PATH("-");

	bool want_stdout(string const &output) {
		return output == STDOUT_PATH;
	}

	int invalidate(string const &dev, string const &output, flags const &fs) {
		try {
			set<uint32_t> blocks;
			block_manager<>::ptr bm = open_bm(dev, block_manager<>::READ_ONLY, !fs.metadata_snapshot_);

			if (fs.metadata_snapshot_) {
				superblock sb = read_superblock(bm);
				if (!sb.metadata_snap)
					throw runtime_error("no metadata snapshot taken.");

				metadata::ptr md(new metadata(bm, *sb.metadata_snap));
				mark_blocks_since(*md, fs.era_threshold_, blocks);

			} else {
				metadata::ptr md(new metadata(bm, metadata::OPEN));
				mark_blocks_since(*md, fs.era_threshold_, blocks);
			}

			if (want_stdout(output))
				emit_blocks(cout, blocks);

			else {
				ofstream out(output.c_str());
				emit_blocks(out, blocks);
			}

		} catch (std::exception &e) {
			cerr << e.what() << endl;
			return 1;
		}

		return 0;
	}
}

//----------------------------------------------------------------

era_invalidate_cmd::era_invalidate_cmd()
	: command("era_invalidate")
{
}

void
era_invalidate_cmd::usage(std::ostream &out) const
{
	out << "Usage: " << get_name() << " [options] --written-since <era> {device|file}\n"
	    << "Options:\n"
	    << "  {-h|--help}\n"
	    << "  {-o <xml file>}\n"
	    << "  {--metadata-snapshot}\n"
	    << "  {--written-since}\n"
	    << "  {-V|--version}" << endl;
}

int
era_invalidate_cmd::run(int argc, char **argv)
{
	int c;
	flags fs;
	string output("-");
	char const shortopts[] = "ho:V";

	option const longopts[] = {
		{ "help", no_argument, NULL, 'h' },
		{ "output", required_argument, NULL, 'o' },
		{ "version", no_argument, NULL, 'V' },
		{ "metadata-snapshot", no_argument, NULL, 1},
		{ "written-since", required_argument, NULL, 2},
		{ NULL, no_argument, NULL, 0 }
	};

	while ((c = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
		switch(c) {
		case 1:
			fs.metadata_snapshot_ = true;
			break;

		case 2:
			fs.era_threshold_ = lexical_cast<uint32_t>(optarg);
			break;

		case 'h':
			usage(cout);
			return 0;

		case 'o':
			output = optarg;
			break;

		case 'V':
			cout << THIN_PROVISIONING_TOOLS_VERSION << endl;
			return 0;

		default:
			usage(cerr);
			return 1;
		}
	}

	if (argc == optind) {
		cerr << "No input file provided." << endl;
		usage(cerr);
		return 1;
	}

	if (!fs.era_threshold_) {
		cerr << "Please specify --written-since" << endl;
		usage(cerr);
		return 1;
	}

	return invalidate(argv[optind], output, fs);
}

//----------------------------------------------------------------
